funsor/cnf.py

```diff
@@ -0,0 +1,340 @@
+        self.red_op = red_op
+        self.bin_op = bin_op
+        self.terms = terms
+        self.reduced_vars = reduced_vars
+        self.is_affine = self._is_affine()
+
+    def _is_affine(self):
+        for t in self.terms:
+            if not isinstance(t, (Number, Tensor, Variable, Contraction)):
+                return False
+            if isinstance(t, Contraction):
+                if not ((t.bin_op, self.bin_op) in DISTRIBUTIVE_OPS or (self.bin_op, t.bin_op) in DISTRIBUTIVE_OPS) \
+                        and t.is_affine:
+                    return False
+
+        if self.bin_op is ops.add and self.red_op is not nullop:
+            return sum(1 for k, v in self.inputs.items() if v.dtype == 'real') == \
+                sum(sum(1 for k, v in t.inputs.items() if v.dtype == 'real') for t in self.terms)
+        return True
+
+    def unscaled_sample(self, sampled_vars, sample_inputs):
+        if self.red_op in (ops.logaddexp, nullop) and self.bin_op in (ops.logaddexp, ops.add, nullop):
+            new_terms = tuple(v.unscaled_sample(sampled_vars.intersection(v.inputs), sample_inputs) for v in self.terms)
+            return Contraction(self.red_op, self.bin_op, self.reduced_vars, *new_terms)
+        raise TypeError("Cannot sample through ops ({}, {})".format(self.red_op, self.bin_op))
+
+    def align(self, names):
+        assert isinstance(names, tuple)
+        assert all(name in self.inputs for name in names)
+        new_terms = tuple(t.align(tuple(n for n in names if n in t.inputs)) for t in self.terms)
+        result = Contraction(self.red_op, self.bin_op, self.reduced_vars, *new_terms)
+        if not names == tuple(result.inputs):
+            return Align(result, names)  # raise NotImplementedError("TODO align all terms")
+        return result
+
+    def _alpha_convert(self, alpha_subs):
+        reduced_vars = frozenset(alpha_subs.get(k, k) for k in self.reduced_vars)
+        bound_types = {}
+        for term in self.terms:
+            bound_types.update({k: term.inputs[k] for k in self.bound.intersection(term.inputs)})
+        alpha_subs = {k: to_funsor(v, bound_types[k]) for k, v in alpha_subs.items()}
+        red_op, bin_op, _, terms = super()._alpha_convert(alpha_subs)
+        return red_op, bin_op, reduced_vars, terms
+
+
+@recursion_reinterpret.register(Contraction)
+def recursion_reinterpret_contraction(x):
+    return type(x)(*map(recursion_reinterpret, (x.red_op, x.bin_op, x.reduced_vars) + x.terms))
+
+
+@eager.register(Contraction, AssociativeOp, AssociativeOp, frozenset, Variadic[Funsor])
+def eager_contraction_generic_to_tuple(red_op, bin_op, reduced_vars, *terms):
+    return eager(Contraction, red_op, bin_op, reduced_vars, tuple(terms))
+
+
+@eager.register(Contraction, AssociativeOp, AssociativeOp, frozenset, tuple)
+def eager_contraction_generic_recursive(red_op, bin_op, reduced_vars, terms):
+
+    # push down leaf reductions
```

Comment:
Do I understand correctly that this will be unable to reduce variables that occur in multiple factors of an integrand, which is strictly weaker than `@integrator`? Or are they made equivalent via a distributivity rule?

Reply:
They're made equivalent by the second part of this pattern